### PR TITLE
Add CodeQL language

### PR DESCRIPTION
**Pull request by rewinfrey**
This adds Linguist detection support for the [QL](https://help.semmle.com/QL/learn-ql/introduction-to-ql.html) language.

## Checklist:
- [x] **I am adding a new language.**
  - [x] The extension of the new language is used in hundreds of repositories on GitHub.com.

    - Search results for each extension:
      -  [search for .ql](https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3Aql+import+select+from)
      - [search for .qll](https://github.com/search?q=extension%3Aqll+predicate&type=Code)
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - Sample source(s) for `.ql`:
      - https://github.com/tree-sitter/tree-sitter-ql/blob/updated-ql-grammar/examples/002.ql (MIT license)
      - https://github.com/Semmle/ql/blob/bd9a2d71bacfa94cc93ce6ab33dd4c7f8ce71050/csharp/ql/src/Linq/RedundantSelect.ql (Apache License 2.0)
      - https://github.com/krisds/cobol-codeql/blob/9373a45da534889cb5d3ae1841a5233a38abceee/queries/semmlecode-cobol-queries/SQL/SpecifyWhereClauseOnSelect.ql (MIT)
    - Sample source(s) for `.qll`:
      - https://github.com/github/rasmuswl-temp-ql-clone/blob/5e1c8fe8c9c3505c35f7e032169bf56d71ea2bc7/cpp/ql/src/semmle/code/cpp/models/implementations/Inet.qll
      - https://github.com/github/rasmuswl-temp-ql-clone/blob/5e1c8fe8c9c3505c35f7e032169bf56d71ea2bc7/python/ql/src/semmle/python/objects/Classes.qll
      - https://github.com/github/rasmuswl-temp-ql-clone/blob/5e1c8fe8c9c3505c35f7e032169bf56d71ea2bc7/python/ql/src/semmle/python/objects/Descriptors.qll
    - Sample license(s):
      - MIT
      - Apache License 2.0
  - [x] I have included a syntax highlighting grammar: [QL lighthouse example](https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-url&grammar_format=auto&grammar_url=https%3A%2F%2Fgithub.com%2FSemmle%2Fql-syntax%2Fblob%2Fmaster%2Fql.tmLanguage.json&grammar_text=&code_source=from-url&code_url=https%3A%2F%2Fgithub.com%2Fkrisds%2Fcobol-codeql%2Fblob%2F9373a45da534889cb5d3ae1841a5233a38abceee%2Fqueries%2Fsemmlecode-cobol-queries%2FSQL%2FSpecifyWhereClauseOnSelect.ql&code=)

/cc @alexet

### CODE REVIEW

File: lib/linguist/languages.yml [hunk 1]
```diff
@@ -4207,6 +4207,14 @@ Python traceback:
   tm_scope: text.python.traceback
   ace_mode: text
   language_id: 304
+QL:
+  type: programming
+  extensions:
+  - ".ql"
+  - ".qll"
```
Comment (pchaigno) [inline lib/linguist/languages.yml]: We'll need a search query for this extension, as well as a sample file.

Comment (rewinfrey) [reply lib/linguist/languages.yml]: @pchaigno thank you for catching this. My apologies for overlooking that. I've updated the PR body to include a `.qll` search along with `.qll` samples and their sources (the samples provided are licensed under Apache 2.0).

**Comment by rewinfrey** [comment]
> I was also wondering if [these files](https://github.com/search?q=extension%3Aql+form&type=Code) are of the same language?

@pchaigno the first file in that query you provided is a `.ql` file that should be identified as `QL`, but the others are not. `form` is not a keyword in `QL` but can be used as a variable name (as seen in the first `.ql` file in the search query).

File: lib/linguist/languages.yml (hunk 1 above)
Comment (pchaigno) [reply lib/linguist/languages.yml]: Thanks!

**Comment by pchaigno** [comment]
Thanks for the answers and updated pull request!

> @pchaigno the first file in that query you provided is a `.ql` file that should be identified as `QL`, but the others are not.

We'll need to identify the language(s) of those files. If there are enough, we'll need to add support for those languages in Linguist. Otherwise, all those files are going to be identified as QL.

What about [these files](https://github.com/search?utf8=%E2%9C%93&q=extension%3Aqll+predicate+javascript&type=Code)?

**Comment by maxbrunsfeld** [comment]
> What about these files?

All of those files (at least the first 5 pages) *are* QL 👍 .

**Comment by maxbrunsfeld** [comment]
> We'll need to identify the language(s) of those files. If there are enough, we'll need to add support for those languages in Linguist. Otherwise, all those files are going to be identified as QL.

It looks like for those files, the language that has been invented as an assignment for a programming class at a Dutch university.

> The goal of this assignment is to build a DSL for questionnaires, called QL. QL allows you to define simple forms with conditions and computed values.

Source: the readme in https://github.com/EdwinOuwehand/ql-dsl.

I think that these other two repos in your search results are related to the same university course:
https://github.com/software-engineering-amsterdam/multi-ql
https://github.com/software-engineering-amsterdam/endless-ql

So it looks to me that the `QL` language used in this course is not widely used, and it is probably ok for the `.ql` files that correspond to this language to be classified as "QL".

@rewinfrey @pchaigno Thoughts?

**Comment by rewinfrey** [comment]
> What about [these files](https://github.com/search?utf8=%E2%9C%93&q=extension%3Aqll+predicate+javascript&type=Code)?

Confirming along with @maxbrunsfeld that those are `QL` files.

> So it looks to me that the `QL` language used in this course is not widely used, and it is probably ok for the `.ql` files that correspond to this language to be classified as "QL".

That's how I see it, too. The fact that these two separate languages happen to also share the same name is a funny coincidence.

@pchaigno as far as I know, the only official programming language using `.ql` and `.qll` as file extensions is Semmle's `QL`.

**Comment by alexet** [comment]
We need to change the language name here to CodeQL.

**Comment by alexet** [comment]
Sorry I was confused about the naming. This should stay as QL

**Comment by rewinfrey** [comment]
👋 @pchaigno when you have a moment, would you mind taking another pass over this? I think the concerns about other `.ql` files have been addressed, but please let me know if I've missed something. 🙇

**Comment by rewinfrey** [comment]
@alexet confirming I've updated Linguist in this patch to use https://github.com/github/vscode-codeql as the definitive grammar in https://github.com/github/linguist/pull/4805/commits/efe19e8321f34f5a67e81b690641ca086207fc06

**Comment by rewinfrey** [comment]
A decision was made to update the name from `QL` to `CodeQL`, I've updated this patch to reflect that.

@pchaigno friendly ping 😄 This is planned to be deployed to github.com as soon as this patch is ready.

/cc @alexet

**Comment by pchaigno** [comment]
> It looks like for those files, the language that has been invented as an assignment for a programming class at a Dutch university.
> 
> > The goal of this assignment is to build a DSL for questionnaires, called QL. QL allows you to define simple forms with conditions and computed values.
> 
> Source: the readme in https://github.com/EdwinOuwehand/ql-dsl.
> 
> I think that these other two repos in your search results are related to the same university course:
> https://github.com/software-engineering-amsterdam/multi-ql
> https://github.com/software-engineering-amsterdam/endless-ql
> 
> So it looks to me that the `QL` language used in this course is not widely used, and it is probably ok for the `.ql` files that correspond to this language to be classified as "QL".

@maxbrunsfeld Thanks for digging into this :bowing_man: 

LGTM otherwise!

**Pull request merged**